### ISSUE TITLE
Audio during streaming is laggy

### ISSUE THREAD

**Issue by kboskin** [pedroSG94/rtmp-rtsp-stream-client-java#277]
Hi! Thanks for great library for streaming. During implementation in my project i faced an issue that sound is a bit laggy. Here is my configuration

 private val AUDIO_BITRATE: Int = 2500 * 1024
 private val SAMPLE_RATE: Int = 44100

and here is the camera

rtmpCamera1.prepareAudio(AUDIO_BITRATE, SAMPLE_RATE, true, true, true)

Maybe you can help to figure out? Thanks!

**Comment by kostyaBoss** [comment]
No, I am streaming to amazon medialive

**Comment by pedroSG94** [comment]
I don't know reason, for now, try emulate YouTube fix. Add delay to server and we can discard it (around 30s). Also try use multiple sample rate 32000, 22050, 16000, 11025.

Other cause could be audio profile. Can you tell me profile used by OBS? I'm using LC (Low Complexity) because all Android device support this profile.
I will try find a solution using YouTube because fix it could fix your problem too.

**Comment by mhackeras** [comment]
I also wanted to issue this again. Cause I tried your solution pedro with a lot of sample rates but audio was laggy. But when video is processed and saved on YouTube audio is fine... 🤔

**Comment by mhackeras** [comment]
Also i should note here that months before i didnt have this problem...

**Comment by pedroSG94** [comment]
If I select normal latency the problem is solved in YouTube but with low or ultra low latency audio is flickering.

For now, I tried modify packetizer, use AAC HE and use AAC HEv2 without luck :(.

If you can remember library version used that could be helpful to find the error.

**Comment by kostyaBoss** [comment]
@pedroSG94 Today used youtube for streaming with low latency - sound was flickering. But I am most of time streaming using amazon medialive service. Btw, I've noticed, that my stream's dashboard was showing 18-22 FPS most of time during streaming, but video was in good quality. Not sure if this is somehow connected with sound

**Comment by pedroSG94** [comment]
FPS is not related with this audio error. The problem is with YouTube and Amazon medialive because if you use other platform all is working fine (twitch, facebook, wowza, etc).
I think that some configuration in rtmp audio packetizer is not supported by YouTube and Amazon medialive.

**Comment by kostyaBoss** [comment]
@pedroSG94 Ok, I will make more detailed discovery tomorrow. Version of library I am currently using is 1.4.9

**Comment by kostyaBoss** [comment]
@mhackeras can you share please version of library, in which this kind of situation wasn't reproducing?

**Comment by mhackeras** [comment]
It was before fpslimiter pedro might remember I can't remember the version.
But I tried without fpslimiter and had the same problem. So... I did some change in code but didn't find the time to try. I will reply in some hours from now when I ll test.

**Comment by mhackeras** [comment]
Also I ll note something again.. when streaming audio is indeed laggy.. but when video is saved on youtube and processed.. audio is fine.

**Comment by KostyaBoskin** [comment]
Tested right now with twitch service, everything was good

**Comment by mhackeras** [comment]
As it is ?

**Comment by KostyaBoskin** [comment]
Sound is in good quality as Pedro mentioned

**Comment by kostyaBoss** [comment]
@pedroSG94 I was making a little research and found that Larix Broadcaster is streaming smoothly. Right now trying to understand what is the difference, currently didn't find any reason why there is no flickering

**Comment by ninyesses** [comment]
I have the same problem.
https://www.youtube.com/watch?v=KMCMbSaVsdM
00:00 ~ 00:30 mic enable
00:30 ~ 01:06 mic disable
You can hear the constant noise the microphone makes even when it’s disabled.

**Comment by ninyesses** [comment]
@kostyaBoss 
sample rate : 44100
bitrate : 128k
channel : stereo (but mono has same issue)
audio post process : all off

**Comment by kostyaBoss** [comment]
@ninyesses I mean on youtube side. My stream on youtube was really laggy, but at stream you attached as link I can't hear flickering

**Comment by ninyesses** [comment]
@kostyaBoss Try increasing the volume.
disable DVR, low latency
I have not set anything else apart

**Comment by kostyaBoss** [comment]
@ninyesses have you tried to stream directly on Amazon MediaLive?

**Comment by pedroSG94** [comment]
For now, I'm testing emulating ffmpeg packetizer without result. I will try other more things.
Let me know if you have any idea to test it.

**Comment by pedroSG94** [comment]
Finally I solved the issue with my Samsung S7 in Youtube ultra low latency.
The problem was AAC encoder because my library select google software encoder. I don't know the reason but Youtube doesnt like it.
You can force hardware audio encoder with this line (You should write it in any time before prepareAudio method, for example after create RtmpCamera1 instance like me):
```java
rtmpCamera1 = new RtmpCamera1(openGlView, this);
//This line fix my problem
rtmpCamera1.setForce(CodecUtil.Force.FIRST_COMPATIBLE_FOUND, CodecUtil.Force.HARDWARE);
```
I will priority hardware audio codec in FIRST_COMPATIBLE_FOUND mode to fix this error.
Let me know if this fix your error. If my solution no work for you at least we know that error is audio encoder configuration.

**Comment by kostyaBoss** [comment]
@pedroSG94 Ok, for me it gave an error E/AudioEncoder: Valid encoder not found. And refused to start stream. Here is stacktrace 
![image](https://user-images.githubusercontent.com/20925367/51024687-128f1280-153f-11e9-996d-10f625552734.png)

My environment is Xiaomi Redmi 3s PRO

**Comment by KostyaBoskin** [comment]
@pedroSG94 need some time, I am using library as implementation in my project:)

**Comment by kostyaBoss** [comment]
@pedroSG94 I am using library as implementation, need some time:)

**Comment by kostyaBoss** [comment]
@pedroSG94 tested solution on galaxy sound is perfect. Is there some way to handle the situation on devices with incorrect encoders?

**Comment by pedroSG94** [comment]
I dont know. I will try find a solution with omx.google.aac.encoder but for now no results

**Comment by mhackeras** [comment]
so shouldnt we use the solution you already pushed in project? to avoid incompatibilities?

**Comment by pedroSG94** [comment]
Yes, you should use that commit. But it is only selecting other encoder if you have more than omx.google.aac.encoder. So some devices will not work if you only have google audio encoder.

**Comment by mhackeras** [comment]
then add a check.. try to get other encoder.. if not exist turn back to google's encoder. Avoid it.

**Comment by pedroSG94** [comment]
I did that:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/audio/AudioEncoder.java#L190

**Comment by kostyaBoss** [comment]
@pedroSG94 if not setting up force with your solution on Samsung Galaxy S8 it doesnt streams but only prints frame discarded

**Comment by pedroSG94** [comment]
Can you use this code on S8 and show me the result?
```java
private void showAudioEncoder() {
    List<MediaCodecInfo> mediaCodecInfoList = CodecUtil.getAllEncoders(CodecUtil.AAC_MIME);
    for (MediaCodecInfo mediaCodecInfo : mediaCodecInfoList) {
      String name = mediaCodecInfo.getName();
      Log.e("Pedro", name);
    }
  }
```

**Comment by kostyaBoss** [comment]
@pedroSG94 frame discard was due to nstable internet connection. Right now with new approach i started to suffer from bug that sometimes video is not being translated (in logs only RTMP audio packets). About you questions here are audio encoders 

2019-01-15 14:06:51.799 20574-20574/stage.point.video.com.pointmobile D/MyTagCodecs: OMX.google.aac.encoder
2019-01-15 14:06:51.799 20574-20574/stage.point.video.com.pointmobile D/MyTagCodecs: [audio/mp4a-latm]
2019-01-15 14:06:51.799 20574-20574/stage.point.video.com.pointmobile D/MyTagCodecs: OMX.SEC.naac.enc
2019-01-15 14:06:51.799 20574-20574/stage.point.video.com.pointmobile D/MyTagCodecs: [audio/mp4a-latm]

**Comment by kostyaBoss** [comment]
@pedroSG94 it seems that after frame discard (on samsung s8) when preview is not ready sometimes video is not recovered when preview is started

**Comment by pedroSG94** [comment]
I think that video error is not related with my last commit because I only changed AudioEncoder class.
You mentioned that you was using gradle library compile. Can you tell me version used?

**Comment by kostyaBoss** [comment]
@pedroSG94 version from your master branch with latest updates, I ve already reintegrated modules from your project in mine in debug purposes and encoders discovery, so I got your latest updates. Video issue is really not related to this I think, just reported about that here

**Comment by kostyaBoss** [comment]
@pedroSG94 right now I am looking for some way of custom encoder integration

**Comment by pedroSG94** [comment]
> @pedroSG94 it seems that after frame discard (on samsung s8) when preview is not ready sometimes video is not recovered when preview is started

I will try reproduce it. Can you tell me how to do it?

Do you means pause and restart activity?. For example block device and unblock while stream. In this case you should stop stream in surfaceDestroyed and start in surfaceChanged as you can see in activities samples. You should start stream or preview after surfaceChanged callback because you need wait that device inflate view.

**Comment by kostyaBoss** [comment]
@pedroSG94 lets create for this another issue but here leave conversation only for encoder troubles

**Comment by mhackeras** [comment]
Are frames discarded or you didn't setup an animation on notification so as the screen to refresh?

**Comment by kostyaBoss** [comment]
@pedroSG94 how about try to compile in project [this lib?](https://github.com/USBhost/MX_FFmpeg) I am reproducing issue with audio on Nexus devices:( Or maybe you have some ideas about how to compile with project another encoder?

**Comment by pedroSG94** [comment]
Compile ffmpeg is not useful in this case. Maybe you can replace mediacodec with ffmpeg audio encoder but I'm not familiar with C++ to write a wrapper for it, for this reason I discarded this idea. You should use this library of ffmpeg:
https://github.com/FFmpeg/FFmpeg/tree/master/libavcodec

I already tried with all mediacodec configurations supported with google audio encoder but no luck :(.
The best way (I think) is try find buffer differences with google audio encoder and other encoders and try fix the error manually (This is hard because I don't know AAC encoding algorithm).

I haven't more ideas for now. I will notify here if I find other way.

**Comment by kostyaBoss** [comment]
@pedroSG94 thanks for your answer. Please don't put this issue away, I think it is very important one for your lib users:grin:

Will make some research in direction you provided and put updates here

**Comment by mhackeras** [comment]
Guys any news here?
i used latest solution but audio is laggy again.. :S cant find a solution on this :/

**Comment by pedroSG94** [comment]
With YouTube and AWS if you are using low latency with google audio encoder you will experiment this error. The only solution in YouTube is use normal latency and maybe you can do a similar solution for AWS. I don't know if I can fix it in the library. I tried but no luck for now.
I can do other try but I can't promise you anything.

**Comment by mhackeras** [comment]
@pedroSG94 live stream on youtube is bad with whatever setting i use.. on encoders, on latency for youtube and bitrates/samplerates.
the worst thing is that youtube doesnt have any video analyzer.. and i cant tell what the problem is..

**Comment by pedroSG94** [comment]
I means that you should select on youtube normal latency option when you create a event:
![48853356-9e8e8480-edaf-11e8-82d8-9141527f52bb](https://user-images.githubusercontent.com/17556876/55908994-2e0bc700-5bdb-11e9-9f52-84d9d633a552.png)
Obviously it isn't a solution because this produce 30s delay but I only find this.

**Comment by mhackeras** [comment]
Yeah even with this it won't work good. Audio still laggy.. 😕 I have been searching since yesterday

**Comment by pedroSG94** [comment]
S7 with hardware encoder is working for me. Did you enabled noise suppressor or echo cancelator? I tested with both disable (Basically default prepareAudio method).
Share me audio and video configuration. Mode used(Display, Camera2,) and view used to test. Also YouTube event configuration.

**Comment by mhackeras** [comment]
I have tested with every settings.
But ok I lol give you some settings that got to work some time ago.
Display.rtmp with opengl
Samplerate - 44100
Bitrates - 128kbps
No audio modes

Video
720p
3500k bitrates
30fps
Youtube I tested With every mode generally normal latency.

**Comment by pedroSG94** [comment]
This work for you in old library version? In this case. Can you tell me a version working to see differences?
If you only remember a date you can find versions with upload date here:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/releases
Anyway, I will test tomorrow and report you the result. Also if you can tell me if audio is good when you save the stream.

**Comment by mhackeras** [comment]
Yeah audio is good when recording.. and is good when I stop stream and
video is saved on youtube.
I stream with some settings on twitch.. and it's super good and with very
low latency.
I cant really remember but it was before fps limiter.. though I tested
without fps limiter and no good result.

**Comment by pedroSG94** [comment]
I will test with a version with fpslimiter according with commits 1.5.6 should have fps limiter

**Comment by mhackeras** [comment]
Also we had made some changes to fix mute audio etc. On microphone manager.
After these I noticed the bug.. before this I streamed to youtube without a
problem.

**Comment by mhackeras** [comment]
Don't misjudge me pedro.. but without your help I don't think I can find the reason. Been searching for hours! And gonna keep looking. It's a pain in t @@s. 😕

**Comment by pedroSG94** [comment]
This is fine. I think this is the bigger bug in the library and I really want fix it also get help to debug and find the solution is great :D

**Comment by mhackeras** [comment]
Hehe I mean I don't expect from you more. You have done the best job here..
and this is another real challenge  I am running tons of tests since
yesterday

**Comment by kostyaBoss** [comment]
@pedroSG94 main reason for working on Samsung devices is replacing OMX with custom one's. When I performed discovery phase I've noticed that [those guys](https://softvelum.com/larix/) created solution which makes stream on AWS really smooth. I've also tried to integrate custom audio codecs into project but no luck, haven't enough knowledges i native part. Also I've tried lot of solutions from internet and all of them had this issue. I think the problem even not in your solution, @pedroSG94 but somewhere deeper

**Comment by pedroSG94** [comment]
I'm testing with version 1.5.6 and audio look fine in S7 hardware audio encoder.and fpslimiter. Maybe I can find a solution for google AAC encoder by this way.

**Comment by mhackeras** [comment]
so on latest version you get bad audio too right?
also yesterday by searching for errors on video with ffmpeg i got this
"non monotonically increasing dts to muxer in stream"
now i am gonna run a full test on video and audio.. timestamps etc.
Please pedro check the timestamps of audio sent too.

**Comment by pedroSG94** [comment]
> so on latest version you get bad audio too right?
> also yesterday by searching for errors on video with ffmpeg i got this
> "non monotonically increasing dts to muxer in stream" and i got a lot of these!
> when i placed fps limiter on video encoder again i didnt get any of this error again.
> 
> now i am gonna run a full test on video and audio.. timestamps etc.
> Please pedro check the timestamps of audio sent too.

You are right if you delete this lines audio work fine with hardware encoder and without fpslimiter ([I don't know if this affect to twitch sync](https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/issues/303)):
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/audio/AudioEncoder.java#L148
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/audio/AudioEncoder.java#L177
I'm testing timestamps but no luck with google encoder.

**Comment by mhackeras** [comment]
maybe these lines arent needed too.?
i mean fpslimiter on video encoder may produced the non synced audio.
I mean pedro what could cause a distorted audio?

**Comment by mhackeras** [comment]
Also confirmed without these lines and no limit on encoder everything working fine!
i gonna test google encoder too.
this is so .... :D eehh

**Comment by mhackeras** [comment]
Also audio is good synced with video.. tested on youtube!

**Comment by mhackeras** [comment]
if this helps.. i can hear something i said 2 secs before or more.. like an echo. when using the bugged code.

**Comment by mhackeras** [comment]
take a look here
https://stackoverflow.com/questions/18857692/muxing-aac-audio-with-androids-mediacodec-and-mediamuxer/18966374#18966374

**Comment by mhackeras** [comment]
as i can see everywhere for timestamps they calculate it diffirently.. 
like here 👍 https://stackoverflow.com/questions/10025824/how-to-record-audio-voice-in-background-continuously-in-android

**Comment by pedroSG94** [comment]
> take a look here
> https://stackoverflow.com/questions/18857692/muxing-aac-audio-with-androids-mediacodec-and-mediamuxer/18966374#18966374

Tested that but no results.

**Comment by mhackeras** [comment]
check latest link i sent. :/

**Comment by mhackeras** [comment]
everywhere i see it seems to be essential to count the buffersize.
do you think is this related?

**Comment by pedroSG94** [comment]
I don't know :(
I tested your last link but no result. I will continue with timestamp this look the reason because if you change timestamp on AudioEncoder this produce a similar effect to use google encoder

**Comment by mhackeras** [comment]
Yeap. At some hours from now i will have free time to test more.
really a pain :S

**Comment by mhackeras** [comment]
without these lines ... audio is good with hardware encoder and with google.. almost good.
not a super big bug.. i believe today will have it fixed

**Comment by mhackeras** [comment]
Check also the adts header. Will be back on some hours from now

**Comment by mhackeras** [comment]
presentationTimeUs
Added in API level 16
public long presentationTimeUs

The presentation timestamp in microseconds for the buffer. This is derived from the presentation timestamp passed in with the corresponding input buffer. This should be ignored for a 0-sized buffer.

https://developer.android.com/reference/android/media/MediaCodec.BufferInfo

**Comment by mhackeras** [comment]
http://danlec.com/st4k#questions/47165734

tried to place audio according to this but cant seem to make it work.. :S

**Comment by pedroSG94** [comment]
I already tried with adts but no luck. I will check both links now.

**Comment by mhackeras** [comment]
https://stackoverflow.com/questions/18857692/muxing-aac-audio-with-androids-mediacodec-and-mediamuxer/18966374#18966374

i ll try this again.. audio timestamps can be really buggy. gonna try calculate them and see the result.
Though i have to say i tried some things.. but google encoder seems to not cooperate with me :P

**Comment by kostyaBoss** [comment]
Guys, let me know, if I could be useful:)

**Comment by mhackeras** [comment]
after searching more i see that the audioRecord.read() you use pedro needs to be
public int read (byte[] audioData, 
                int offsetInBytes, 
                int sizeInBytes)

Reads audio data from the audio hardware for recording into a byte array. The format specified in the AudioRecord constructor should be AudioFormat#ENCODING_PCM_8BIT to correspond to the data in the array.

From here :
https://developer.android.com/reference/android/media/AudioRecord.html#getMinBufferSize(int,%20int,%20int)

**Comment by pedroSG94** [comment]
I will try change to short but this will be more slow because I need convert it to byte if I want send to encoder (Maybe I can send to encoder with short[] directly but I'm not sure I need investigate it).
Anyway you can read in Android documentation that byte[] read method support PCM16bit but it is deprecated:
https://developer.android.com/reference/android/media/AudioRecord.html#read(byte[],%20int,%20int,%20int)

**Comment by mhackeras** [comment]
Yeap a pain to change all these.
Give me some minutes before you start I am gonna try a method to check the theory that this method is the problem.
Also I read somewhere that converting from short to byte requires 250ms so.. 
Let me check if this is the problem first

**Comment by mhackeras** [comment]
If you remember i said older versions audio was fine with google encoder.. maybe it was the older version of android and encoder that worked.. and changed now. so thats why its deprecated maybe

**Comment by mhackeras** [comment]
Pedro how is it going?
If you have free time cause I dont have at the moment.. try this.
Take timestamp for audio using surface timestamp.

**Comment by pedroSG94** [comment]
I had not time the last 2 days and today  I will take holydays in 2 days threfore I will have more time. 
I have short[] adaptation half way. I will try surface timestamp too.

**Comment by pedroSG94** [comment]
Hi,
I already tested both suggestions (short[] and surface timestamp).
The first idea result in a bad audio quality (worst than use byte[]). The second produce that stream never start. This last is really easy to test you only need add this line to getAacData callback in Camera1Base and use openglview:
```java
info.presentationTimeUs = glInterface.getSurfaceTexture().getTimestamp();
```

**Comment by mhackeras** [comment]
Yeap i understand.
I will check this with the timestamps.

**Comment by mhackeras** [comment]
Pedro hi how are you?
recently after i couldnt start audio record on oneplus 6 with default source.. found out that should use noise suppression so as to start..
so was searching on this and found that some phones produce bad audio like what is happening with google encoder.. so maybe try with another audio source or something?

**Comment by pedroSG94** [comment]
I already tried with all sources of this list:
https://developer.android.com/reference/android/media/MediaRecorder.AudioSource

I did a test with fromfile mode. My idea was see buffer difference from a file that should have same audio buffer (I used a video of 1second) using a encoder that work (A hardware encoder by Smasung) and google encoder. The result was that it is totally different so I think that google encoder use a different algorithm.

Zip with files:

[testbuffers.zip](https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/files/3171753/testbuffers.zip)

**Comment by mhackeras** [comment]
Hi pedro.
after doing some work on audio.. i am now pretty sure that we should save the audio to short.
I may have a solution today with google encoder but i will text here.
I wanted to ask you if you have any audio filters.. i may do some to import to your project.

**Comment by pedroSG94** [comment]
Hi,
I have this audio filters from Android api:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/encoder/src/main/java/com/pedro/encoder/input/audio/AudioPostProcessEffect.java
Official documentation:
https://developer.android.com/reference/android/media/audiofx/AudioEffect.html

**Comment by mhackeras** [comment]
I mean custom filters.. by manipulating shorts from audio pcm.
Will talk soon

**Comment by mhackeras** [comment]
@pedroSG94 when using google encoder setting timestamp on audioinfo is essential cause sometimes new buffer has lower millis than last buffer.
instead of setting timestamps which causes bad audio on streams you can check if current timestamp is bigger than last and write otherwise drop.
though this doesnt yet fix the audio lag when streaming using omx.google.aac

**Comment by mhackeras** [comment]
also i found good practice settings buffer size using getMinPcmbuffer and also set this in max input size on audio encoder. This makes audio recording more stable and compatible.
instead of using fixed buffer size..

**Comment by pedroSG94** [comment]
Hi,
Good suggestions. I will implement this changes.

**Comment by mhackeras** [comment]
@pedroSG94  there are times that i get trackIndex is invalid. Cant reproduce in standard use.

 java.lang.IllegalArgumentException: trackIndex is invalid
        at android.media.MediaMuxer.writeSampleData(MediaMuxer.java:657)
        at com.pedro.rtplibrary.base.RecordController.recordAudio(RecordController.java:143)
        at com.pedro.rtplibrary.base.DisplayBase.getAacData(DisplayBase.java:462)
        at com.pedro.encoder.audio.AudioEncoder.getDataFromEncoderAPI21(AudioEncoder.java:173)
        at com.pedro.encoder.audio.AudioEncoder.inputPCMData(AudioEncoder.java:143)
        at com.pedro.rtplibrary.base.DisplayBase.inputPCMData(DisplayBase.java:490)
        at com.pedro.encoder.input.audio.MicrophoneManager$1.run(MicrophoneManager.java:164)
        at java.lang.Thread.run(Thread.java:764)

**Comment by mhackeras** [comment]
the above happens when i prepare audio and video and start record after some seconds.
Displayrtmp

**Comment by pedroSG94** [comment]
Ok. I will try fix it

**Comment by mhackeras** [comment]
@pedroSG94 dont think i expect you to fix things.. i work on your project every day.
but this error is so strange.. sound timestamps are buggy for some reason i cant understand :/

**Comment by mhackeras** [comment]
It seems that if record doesnt start very soon audio timestamps are wrong. But how?? Only mediaprojection gets data still start of record. 🤔

**Comment by pedroSG94** [comment]
I tried to reproduce your error recording but I can't do it.
I tried more than 100 times recording in displayrtmp demo but never crash. Is this error really difficult to reproduce? I will try more times and with other devices.

**Comment by mhackeras** [comment]
Yeah it seems difficult to reproduce. It happens when I prepare video and
audio and start recording after a while. Like 3 seconds or more.

**Comment by mhackeras** [comment]
@pedroSG94 i am really sorry it was caused by my implementation for recording without audio :/ i am really sorry!

**Comment by pedroSG94** [comment]
No problem. I'm glad that you find the error.
About the others suggestions.
- I can't reproduce audio timestamp error (I tried with 3 stream of 1h but all timestamp was correct).
- Set max audio buffer to encoder make no different in audio quality, I did a test hardcoding it. This is suppose to be a good practique, so I will try implement it anyway.
- Min audio buffer no tested yet. I'm on it.

**Comment by mhackeras** [comment]
Ok.
1. Timestamp error reproduce:
Set encoder to google encoder and start record.. when streaming there is no real problem.. but when muxing, bad audio timestamp can bring error and stop the record.

2. Setting max audio buffer helps avoid error on diffirent minbuffer for diffirent phones. Some phones have 3000 min buff others 7000. So if you try set 7000 min buff with google encoder you will get error because as a default it has 4096 i think. By setting max input size in encoder it can accept the bigger value you set.

3. Min audio buffer is good because it makes recording stable for every phone.. as soon as you check with Mono channel.

**Comment by pedroSG94** [comment]
- 1 I will test it.
- 2 if you set min buffer in encoder to 0 that means infinite so you can set all size without problem theoretically.

I did a [commit](https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/commit/561ab6affddd8c6772e96bac5c28314f280d48a2) with audio changes:
- I replaced short[] implementation suggestion with bytebuffer that support all encoded formats.
- Min buffer size was calculated according with this [android cocumentation](https://developer.android.com/reference/android/media/AudioRecord.html#getMinBufferSize(int,%20int,%20int))(read note description about set higher values) and looking similar code on github.
- Max audio buffer was set to 4096 that is max bytebuffer data size (I'm not sure if this is your idea but if you are reading 4096 bytes you can't get a buffer more high than it)

**Comment by mhackeras** [comment]
2. I tested it and It is not infinite when on 0.

**Comment by saurik** [comment]
I was running into some audio "crackling" while streaming to YouTube which I believe to be related to this issue. There were two problems I had to fix: one of them was due to a data alignment problem with ByteBuffer (and I've sent in a crude pull request #365 to fix the problem).

The other issue I ran into was related to presentation timestamps (as alluded to by @mhackeras). I do not have a general purpose pull request fix for this one, as I both hardcoded my local fix (which was more of a proof of concept) and also am not even 100% sure what the best behavior here would be anyway; but, essentially, it isn't safe to just call nanoTime().

When you call nanoTime is only sort of correlated to when the audio packet was actually recorded... it is potentially much later, as there might be some lag in processing for various reasons. The result is that later packets might accidentally be marked as being presented after long gaps or "backwards in time" during the middle of the previous audio packet.

As I'm always using mono 44.1kHz audio, I tried a hack where I "calculate" the presentation timestamp based on how many audio samples I've so far streamed, and together with the arrayOffset() fix I was able to remove all of the audio crackling. I maintain the "current calculated timestamp" and then use that as my base rather than calling nanoTime().

```java
presentTimeUsCalculated += (long) size / (long) 2 * (long) 1000000 / (long) 44100;
long pts = presentTimeUsCalculated - presentTimeUs;
```

However, this fix isn't right right, as if there is a buffer overrun due to unexpected processing lag then the presentation timestamp will fail to advance and lead to an audio desynchronization issue. This should be fixed by doing something like _also_ calling nanoTime and seeing if the packet is in the future, and if it is then advancing the presentation timestamp to that point.
